docs/proposals/scheduling/20220629-schedule-device-in-card-level.md

```diff
@@ -0,0 +1,268 @@
+best placement for ci\gi-create and schedule pod in it.
+
+#### Story Oversold
+in some cases, we find although some cards free is zero in scheduler account, but we observed that the physical
+utilization of gpu-mem\gpu-core still has buffer, so we may schedule some low-priority jobs to use this buffer, and
+when buffer disappear, we should kill these jobs to avoid affecting normal jobs. because the utilization is different
+bewteen cards, so scheduler must schedule based card level.
+     
+
+## Implementation Details
+
+#### Scheduler
+target:
+1. we need a new data structure to describe each device card resources and healthy status on one node
+2. we need new plugins to execute filter\reserve\bind stage
+3. support user to apply any ratio of one card, like 50%/25%
+4. automatically recognize different kind devices. when a new device added, we don't need modify any code
+5. schedule-result will be written on pod annotations.
+
+detail:
+
+1.device resoruce dimension:
+For GPU, we have 3 dimensions:gpu-core-ratio\gpu-mem-ratio\gpu-mem, this design is for future node isolation feature. 
+
+   gpu-core-ratio means calculation percentage of one card.
+
+   gpu-mem-ratio means memory percentage of one card. 
+
+   gpu-mem is real memory-byte of pod usage. 
+
+   we assume one card's total resource is {gpu-core-ratio:100, gpu-mem-ratio:100, gpu-mem:xxGB}, xxGB is reported by device 
+reporter. For other devices like RDMA\FPGA, we temporary only has rdma\fpga demension, total resource is 
+{rdma:100}, {fpga:100}.
+
+2.user apply device resource:
+
+   a. if user apply {nvidia.com/gpu:1}, we translate to {gpu-core-ratio:100, gpu-mem-ratio:100， gpu-mem:0GB}. gpu-mem 0GB
+is because different gpu model has different gpu-memory. so when we filter different nodes, we will determine the 
+real gpu-mem value by node gpu memory * pod's gpu-mem-ratio in real time.
+
+   b. if user apply {koordinator/gpu:100}, we translate to {gpu-core-ratio:100, gpu-mem-ratio:100， gpu-mem:0GB}.
+
+   c. if user apply {koordinator/gpu-core:50, koordinator/gpu-mem-ratio:50}, we translate to {gpu-core-ratio:50, 
+gpu-mem-ratio:50， gpu-mem:0GB}
+
+   d. if user apply {koordinator/gpu-core:50, koordinator/gpu-mem:xxGB}, we translate to {gpu-core-ratio:50, 
+gpu-mem-ratio:0， gpu-mem:XXGB}. when we filter different nodes, we cal gpu-mem-ratio by xxGB / node's total gpu-memory.
+
+   e. when dimension's value > 100, means pod need multi-devices. now we only allow value can be divided by 100.
+
+3.device resource data structure:
+for each device, we use DeviceResource to present resources. DeviceResource's DeviceKey2ValueMap describe each minor's resources
+ ```go
+// "0": {gpu-core-ratio:100, gpu-mem-ratio:100, gpu-mem: 16GB}
+// "1": {gpu-core-ratio:100, gpu-mem-ratio:100, gpu-mem: 16GB}
+type DeviceResource struct {
+    // key is minor 
+    DeviceKey2ValueMap map[int32]map[string]resource.Quantity
```

Comment:
Do not use this XXX2YYY format for naming. And this data structure seems to be only used by scheduling plugins, it is recommended to describe in plugin related details.